File: Geolocation sample/VB/Scenario4UIHandlers.vb
'*********************************************************
'
' Copyright (c) Microsoft. All rights reserved.
' THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
' ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
' IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
' PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
'
'*********************************************************

Imports System.Threading
Imports Windows.Devices.Enumeration
Imports Windows.Devices.Geolocation
Imports Windows.Devices.Geolocation.Geofencing
Imports Windows.Data.Json

''' <summary>
''' An empty page that can be used on its own or navigated to within a Frame.
''' </summary>
Partial Public NotInheritable Class Scenario4
    Inherits SDKTemplate.Common.LayoutAwarePage

    Private Sub OnRadiusTextChanged(ByVal sender As Object, ByVal e As TextChangedEventArgs)
        radiusSet = TextChangedHandlerDouble(False, "Radius", Radius)

        DetermineCreateGeofenceButtonEnableState()
    End Sub

    Private Sub OnLongitudeTextChanged(ByVal sender As Object, ByVal e As TextChangedEventArgs)
        longitudeSet = TextChangedHandlerDouble(False, "Longitude", Longitude)

        DetermineCreateGeofenceButtonEnableState()
    End Sub

    Private Sub OnLatitudeTextChanged(ByVal sender As Object, ByVal e As TextChangedEventArgs)
        latitudeSet = TextChangedHandlerDouble(False, "Latitude", Latitude)

        DetermineCreateGeofenceButtonEnableState()
    End Sub

    Private Sub OnIdTextChanged(ByVal sender As Object, ByVal e As TextChangedEventArgs)
        ' get number of characters
        Dim characterCount As Integer = Id.Text.Length

        nameSet = (0 <> characterCount)

        CharCount.Text = characterCount.ToString() & " characters"

        DetermineCreateGeofenceButtonEnableState()
    End Sub

    Private Sub OnRegisteredGeofenceListBoxSelectionChanged(ByVal sender As Object, ByVal e As SelectionChangedEventArgs)
        Dim list As IList(Of Object) = e.AddedItems

        If 0 = list.Count Then
            ' disable the remove button
            RemoveGeofenceItem.IsEnabled = False
        Else
            ' enable the remove button
            RemoveGeofenceItem.IsEnabled = True

            ' update controls with the values from this geofence item
            ' get selected item
            Dim item As GeofenceItem = TryCast(RegisteredGeofenceListBox.SelectedItem, GeofenceItem)

            RefreshControlsFromGeofenceItem(item)

            DetermineCreateGeofenceButtonEnableState()

        End If
    End Sub

    Private Sub RefreshControlsFromGeofenceItem(ByVal item As GeofenceItem)
        If Nothing IsNot item Then
            Id.Text = item.Id
            Latitude.Text = item.Latitude.ToString()
            Longitude.Text = item.Longitude.ToString()
            Radius.Text = item.Radius.ToString()
            SingleUse.IsChecked = item.SingleUse

            If 0 <> item.DwellTime.Ticks Then
                DwellTime.Text = item.DwellTime.ToString()
            Else
                DwellTime.Text = ""
            End If

            If 0 <> item.Duration.Ticks Then
                Duration.Text = item.Duration.ToString()
            Else
                Duration.Text = ""
            End If

            If 0 <> item.StartTime.Ticks Then
                Dim dt As DateTimeOffset = item.StartTime

                StartTime.Text = formatterShortDateLongTime.Format(dt)
            Else
                StartTime.Text = ""
            End If

            ' Update flags used to enable Create Geofence button
            OnIdTextChanged(Nothing, Nothing)
            OnLongitudeTextChanged(Nothing, Nothing)
            OnLatitudeTextChanged(Nothing, Nothing)
            OnRadiusTextChanged(Nothing, Nothing)
        End If
    End Sub

    Private Function TextChangedHandlerDouble(ByVal nullAllowed As Boolean, ByVal name As String, ByVal e As TextBox) As Boolean
        Dim valueSet As Boolean = False

        Try
            Dim value As Double = Double.Parse(e.Text)

            valueSet = True

            ' clear out status message
            rootPage.NotifyUser("", NotifyType.StatusMessage)
        Catch e1 As ArgumentNullException
            If False = nullAllowed Then
                If Nothing IsNot name Then
                    rootPage.NotifyUser(name & " needs a value", NotifyType.StatusMessage)
                End If
            End If
        Catch e2 As FormatException
            If Nothing IsNot name Then
                rootPage.NotifyUser(name & " must be a number", NotifyType.StatusMessage)
            End If
        Catch e3 As OverflowException
            If Nothing IsNot name Then
                rootPage.NotifyUser(name & " is out of bounds", NotifyType.StatusMessage)
            End If
        End Try

        Return valueSet
    End Function

    Private Function TextChangedHandlerInt(ByVal nullAllowed As Boolean, ByVal name As String, ByVal e As TextBox) As Boolean
        Dim valueSet As Boolean = False

        Try
            Dim value As Integer = Integer.Parse(e.Text)

            valueSet = True

            ' clear out status message
            rootPage.NotifyUser("", NotifyType.StatusMessage)
        Catch e1 As ArgumentNullException
            If False = nullAllowed Then
                If Nothing IsNot name Then
                    rootPage.NotifyUser(name & " needs a value", NotifyType.StatusMessage)
                End If
            End If
        Catch e2 As FormatException
            If Nothing IsNot name Then
                rootPage.NotifyUser(name & " must be a number", NotifyType.StatusMessage)
            End If
        Catch e3 As OverflowException
            If Nothing IsNot name Then
                rootPage.NotifyUser(name & " is out of bounds", NotifyType.StatusMessage)
            End If
        End Try

        Return valueSet
    End Function

    Private Sub SetStartTimeToNowFunction()
        Try
            calendar.SetToNow()

            Dim dt As DateTimeOffset = calendar.GetDateTime()

            StartTime.Text = formatterShortDateLongTime.Format(dt)
        Catch ex As Exception
            rootPage.NotifyUser(ex.ToString(), NotifyType.ErrorMessage)
        End Try
    End Sub

    Private Function GetTimeStampedMessage(ByVal eventCalled As String) As String
        Dim message As String

        calendar.SetToNow()

        message = eventCalled & " " & formatterLongTime.Format(calendar.GetDateTime())

        Return message
    End Function

    ' are settings available so a geofence can be created?
    Private Function SettingsAvailable() As Boolean
        Dim fSettingsAvailable As Boolean = False

        If True = nameSet AndAlso True = latitudeSet AndAlso True = longitudeSet AndAlso True = radiusSet Then
            ' also need to test if data is good
            fSettingsAvailable = True
        End If

        Return fSettingsAvailable
    End Function

    Private Sub DetermineCreateGeofenceButtonEnableState()
        CreateGeofenceButton.IsEnabled = SettingsAvailable()
    End Sub

    ' add geofence to listbox
    Private Sub AddGeofenceToRegisteredGeofenceListBox(ByVal geofence As Geofence)
        Dim item As New GeofenceItem(geofence)

        ' the registered geofence listbox is data bound
        ' to the collection stored in the root page
        geofenceCollection.Insert(0, item)
    End Sub

    ''' <summary>
    ''' This is the click handler for the 'Set Start Time to Now' button.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub OnSetStartTimeToNow(ByVal sender As Object, ByVal e As RoutedEventArgs)
        SetStartTimeToNowFunction()
    End Sub

    ''' <summary>
    ''' This is the click handler for the 'Set Lat/Long to current position' button.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Async Sub OnSetPositionToHere(ByVal sender As Object, ByVal e As RoutedEventArgs)
        ' save off current cursor and set cursor to a wait cursor
        Dim oldCursor As Windows.UI.Core.CoreCursor = Window.Current.CoreWindow.PointerCursor
        Window.Current.CoreWindow.PointerCursor = New Windows.UI.Core.CoreCursor(CoreCursorType.Wait, 0)
        SetPositionToHereButton.IsEnabled = False
        Latitude.IsEnabled = False
        Longitude.IsEnabled = False

        Try
            ' Get cancellation token
            cts = New CancellationTokenSource()
            Dim token As CancellationToken = cts.Token

            ' Carry out the operation
            Dim pos As Geoposition = Await geolocator.GetGeopositionAsync().AsTask(token)

            Latitude.Text = pos.Coordinate.Point.Position.Latitude.ToString()
            Longitude.Text = pos.Coordinate.Point.Position.Longitude.ToString()

            ' clear status
            rootPage.NotifyUser("", NotifyType.StatusMessage)
        Catch e1 As UnauthorizedAccessException
            If DeviceAccessStatus.DeniedByUser = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the user. Enable access through the settings charm.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.DeniedBySystem = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the system. The administrator of the device must enable location access through the location control panel.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.Unspecified = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by unspecified source. The administrator of the device may need to enable location access through the location control panel, then enable access through the settings charm.", NotifyType.StatusMessage)
            End If
        Catch e2 As TaskCanceledException
            rootPage.NotifyUser("Task canceled", NotifyType.StatusMessage)
        Catch ex As Exception
            ' If there are no location sensors GetGeopositionAsync()
            ' will timeout -- that is acceptable.
            Const WaitTimeoutHResult As Integer = CInt(&H80070102)

            If ex.HResult = WaitTimeoutHResult Then ' WAIT_TIMEOUT
                rootPage.NotifyUser("Operation accessing location sensors timed out. Possibly there are no location sensors.", NotifyType.StatusMessage)
            Else
                rootPage.NotifyUser(ex.ToString(), NotifyType.ErrorMessage)
            End If
        Finally
            cts = Nothing
        End Try

        ' restore cursor and re-enable controls
        Window.Current.CoreWindow.PointerCursor = oldCursor
        SetPositionToHereButton.IsEnabled = True
        Latitude.IsEnabled = True
        Longitude.IsEnabled = True
    End Sub

    ''' <summary>
    ''' This is the click handler for the 'Create Geofence' button.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    Private Sub OnCreateGeofence(ByVal sender As Object, ByVal e As RoutedEventArgs)
        Try
            ' This must be done here because there is no guarantee of 
            ' getting the location consent from a geofence call.
            If False = permissionsChecked Then
                GetGeopositionAsync()
                permissionsChecked = True
            End If

            ' get lat/long/radius, the fence name (fenceKey), 
            ' and other properties from controls,
            ' depending on data in controls for activation time
            ' and duration the appropriate
            ' constructor will be used.
            Dim geofence As Geofence = GenerateGeofence()

            ' Add the geofence to the GeofenceMonitor's
            ' collection of fences
            geofences.Add(geofence)

            ' add geofence to listbox
            AddGeofenceToRegisteredGeofenceListBox(geofence)
        Catch e1 As System.UnauthorizedAccessException
            If DeviceAccessStatus.DeniedByUser = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the user. Enable access through the settings charm.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.DeniedBySystem = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the system. The administrator of the device must enable location access through the location control panel.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.Unspecified = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by unspecified source. The administrator of the device may need to enable location access through the location control panel, then enable access through the settings charm.", NotifyType.StatusMessage)
            End If
        Catch e2 As TaskCanceledException
            rootPage.NotifyUser("Canceled", NotifyType.StatusMessage)
        Catch ex As Exception
            ' GeofenceMonitor failed in adding a geofence
            ' exceptions could be from out of memory, lat/long out of range,
            ' too long a name, not a unique name, specifying an activation
            ' time + duration that is still in the past
            rootPage.NotifyUser(ex.ToString(), NotifyType.ErrorMessage)
        Finally
        End Try
    End Sub

    ''' <summary>
    ''' Helper method to invoke Geolocator.GetGeopositionAsync.
    ''' </summary>
    Private Async Sub GetGeopositionAsync()
        rootPage.NotifyUser("Checking permissions...", NotifyType.StatusMessage)

        inGetPositionAsync = True

        Try
            ' Get cancellation token
            cts = New CancellationTokenSource()
            Dim token As CancellationToken = cts.Token

            ' Carry out the operation
            Await geolocator.GetGeopositionAsync().AsTask(token)

            ' clear status
            rootPage.NotifyUser("", NotifyType.StatusMessage)
        Catch e1 As UnauthorizedAccessException
            If DeviceAccessStatus.DeniedByUser = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the user. Enable access through the settings charm.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.DeniedBySystem = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by the system. The administrator of the device must enable location access through the location control panel.", NotifyType.StatusMessage)
            ElseIf DeviceAccessStatus.Unspecified = accessInfo.CurrentStatus Then
                rootPage.NotifyUser("Location has been disabled by unspecified source. The administrator of the device may need to enable location access through the location control panel, then enable access through the settings charm.", NotifyType.StatusMessage)
            End If
        Catch e2 As TaskCanceledException
            rootPage.NotifyUser("Task canceled", NotifyType.StatusMessage)
        Catch ex As Exception
            ' If there are no location sensors GetGeopositionAsync()
            ' will timeout -- that is acceptable.
            Const WaitTimeoutHResult As Integer = CInt(&H80070102)

            If ex.HResult = WaitTimeoutHResult Then ' WAIT_TIMEOUT
                rootPage.NotifyUser("Operation accessing location sensors timed out. Possibly there are no location sensors.", NotifyType.StatusMessage)
            Else
                rootPage.NotifyUser(ex.ToString(), NotifyType.ErrorMessage)
            End If
        Finally
            cts = Nothing
        End Try

        inGetPositionAsync = False
    End Sub

    Private Sub FillRegisteredGeofenceListBoxWithExistingGeofences()
        For Each geofence As Geofence In geofences
            AddGeofenceToRegisteredGeofenceListBox(geofence)
        Next geofence
    End Sub

    Private Sub FillEventListBoxWithExistingEvents()
        Dim settings = ApplicationData.Current.LocalSettings
        If settings.Values.ContainsKey("ForegroundGeofenceEventCollection") Then
            Dim geofenceEvent As String = settings.Values("ForegroundGeofenceEventCollection").ToString()

            If 0 <> geofenceEvent.Length Then
                Dim events = JsonValue.Parse(geofenceEvent).GetArray()

                ' NOTE: the events are accessed in reverse order
                ' because the events were added to JSON from
                ' newer to older.  AddEventDescription() adds
                ' each new entry to the beginning of the collection.
                For pos As Integer = events.Count - 1 To 0 Step -1
                    Dim element = events.GetStringAt(CUInt(pos))
                    AddEventDescription(element)
                Next pos
            End If

            settings.Values("ForegroundGeofenceEventCollection") = Nothing
        End If
    End Sub

    Private Sub SaveExistingEvents()
        Dim jsonArray As New JsonArray()

        For Each eventDescriptor In eventCollection
            jsonArray.Add(JsonValue.CreateStringValue(eventDescriptor.ToString()))
        Next eventDescriptor

        Dim jsonString As String = jsonArray.Stringify()

        Dim settings = ApplicationData.Current.LocalSettings
        settings.Values("ForegroundGeofenceEventCollection") = jsonString
    End Sub

    Private Sub AddEventDescription(ByVal eventDescription As String)
        If eventCollection.Count = maxEventDescriptors Then
            eventCollection.RemoveAt(maxEventDescriptors - 1)
        End If

        eventCollection.Insert(0, eventDescription)
    End Sub

    Private Enum TimeComponent
        day
        hour
        minute
        second
    End Enum

    Private Function ParseTimeSpan(ByVal field As String, ByVal defaultValue As Integer) As Long
        Dim timeSpanValue As Long = 0
        Dim delimiterChars() As Char = {":"c}
        Dim timeComponents() As String = field.Split(delimiterChars)
        Dim start As Integer = 4 - timeComponents.Length

        If start >= 0 Then
            Dim [loop] As Integer = 0
            Dim index As Integer = start
            Do While [loop] < timeComponents.Length
                Dim tc As TimeComponent = CType(index, TimeComponent)

                Select Case tc
                    Case TimeComponent.day
                        timeSpanValue += CLng(decimalFormatter.ParseInt(timeComponents([loop]))) * secondsPerDay

                    Case TimeComponent.hour
                        timeSpanValue += CLng(decimalFormatter.ParseInt(timeComponents([loop]))) * secondsPerHour

                    Case TimeComponent.minute
                        timeSpanValue += CLng(decimalFormatter.ParseInt(timeComponents([loop]))) * secondsPerMinute

                    Case TimeComponent.second
                        timeSpanValue += CLng(decimalFormatter.ParseInt(timeComponents([loop])))

                    Case Else
                End Select
                [loop] += 1
                index += 1
            Loop
        End If

        If 0 = timeSpanValue Then
            timeSpanValue = defaultValue
        End If

        timeSpanValue *= oneHundredNanosecondsPerSecond

        Return timeSpanValue
    End Function

    Private Class GeofenceItemCollection
        Inherits System.Collections.ObjectModel.ObservableCollection(Of GeofenceItem)

    End Class

    Private Class EventDescriptorCollection
        Inherits System.Collections.ObjectModel.ObservableCollection(Of String)

    End Class
End Class

File: Geolocation sample/JavaScript/js/scenario4UIHandlers.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

// Custom templating function
var RegisteredGeofencesItemTemplate = WinJS.Utilities.markSupportedForProcessing(function RegisteredGeofencesItemTemplate(itemPromise) {
    return itemPromise.then(function (currentItem) {

        // Build ListView Item Container div
        var result = document.createElement("div");
        result.className = "GeofenceListViewItemStyle";

        // Build content body
        var body = document.createElement("div");

        // Display title
        if (currentItem.data) {
            var title = document.createElement("span");
            title.innerText = currentItem.data.id + " (" + currentItem.data.latitude + ", " + currentItem.data.longitude + ", " + currentItem.data.radius + ")";
            body.appendChild(title);
        }

        // put the body into the ListView Item
        result.appendChild(body);

        return result;
    });
});

// Custom templating function
var GeofenceEventsItemTemplate = WinJS.Utilities.markSupportedForProcessing(function GeofenceEventsItemTemplate(itemPromise) {
    return itemPromise.then(function (currentItem) {

        // Build ListView Item Container div
        var result = document.createElement("div");
        result.className = "GeofenceListViewItemStyle";

        // Build content body
        var body = document.createElement("div");

        // Display title
        if (currentItem.data) {
            var title = document.createElement("span");
            title.innerText = currentItem.data;
            body.appendChild(title);
        }

        // put the body into the ListView Item
        result.appendChild(body);

        return result;
    });
});

var itemToRemove;
var itemToRemoveId;
var nameElement;
var charCount;
var latitude;
var longitude;
var radius;
var geofenceSingleUse;
var dwellTimeField;
var durationField;
var startTimeField;
var latitudeSet = false;
var longitudeSet = false;
var radiusSet = false;
var nameSet = false;
var secondsPerMinute = 60;
var secondsPerHour = 60 * secondsPerMinute;
var secondsPerDay = 24 * secondsPerHour;
var millisecondsPerSecond = 1000;
var defaultDwellTimeSeconds = 10;
var useStartTime = false;
var setStartTimeToNow;
var setPositionToHere;
var formatterShortDateLongTime;
var formatterLongTime;
var calendar;
var decimalFormatter;

function setupUI() {
    formatterShortDateLongTime = new Windows.Globalization.DateTimeFormatting.DateTimeFormatter("{month.integer}/{day.integer}/{year.full} {hour.integer}:{minute.integer(2)}:{second.integer(2)}", ["en-US"], "US", Windows.Globalization.CalendarIdentifiers.gregorian, Windows.Globalization.ClockIdentifiers.twentyFourHour);
    formatterLongTime = new Windows.Globalization.DateTimeFormatting.DateTimeFormatter("{hour.integer}:{minute.integer(2)}:{second.integer(2)}", ["en-US"], "US", Windows.Globalization.CalendarIdentifiers.gregorian, Windows.Globalization.ClockIdentifiers.twentyFourHour);
    calendar = new Windows.Globalization.Calendar();
    decimalFormatter = new Windows.Globalization.NumberFormatting.DecimalFormatter();

    document.getElementById("createGeofenceButton").disabled = true;
    document.getElementById("removeGeofenceItem").disabled = true;

    nameElement = document.getElementById("name");
    charCount = document.getElementById("charCount");
    latitude = document.getElementById("latitude");
    longitude = document.getElementById("longitude");
    radius = document.getElementById("radius");
    geofenceSingleUse = document.getElementById("geofenceSingleUse");
    dwellTimeField = document.getElementById("dwellTime");
    durationField = document.getElementById("duration");
    startTimeField = document.getElementById("startTime");
    setStartTimeToNow = document.getElementById("setStartTimeToNowButton");
    setPositionToHere = document.getElementById("setPositionToHereButton");

    nameElement.addEventListener("input", onNameChanged, false);
    latitude.addEventListener("input", onLatitudeChanged, false);
    longitude.addEventListener("input", onLongitudeChanged, false);
    radius.addEventListener("input", onRadiusChanged, false);
    setStartTimeToNow.addEventListener("click", onSetStartTimeToNow, false);
}

function getTimeStampedMessage(eventCalled) {
    calendar.setToNow();

    return eventCalled + " " + formatterLongTime.format(calendar.getDateTime());
}

function setStartTimeToNowFunction() {
    var startTime = new Date(); // with no params the startTime is Now

    startTimeField.value = formatterShortDateLongTime.format(startTime);
}

function onSetStartTimeToNow() {
    setStartTimeToNowFunction();
}

function onNameChanged() {
    // get number of characters
    if (nameElement.value) {
        var count = nameElement.value.length;
        charCount.innerText = count.toString() + " characters";
        if (0 !== count) {
            nameSet = true;
        } else {
            nameSet = false;
        }
    } else {
        charCount.innerText = "0 characters";
        nameSet = false;
    }

    document.getElementById("createGeofenceButton").disabled = !settingsAvailable();
}

function setLatitudeInUI(latitudeArg) {
    latitude.value = latitudeArg;
    onLatitudeChanged();
}

function setLongitudeInUI(longitudeArg) {
    longitude.value = longitudeArg;
    onLongitudeChanged();
}

function onLatitudeChanged() {
    if (textChangedHandlerDouble(false, "Latitude", latitude)) {
        latitudeSet = true;
    } else {
        latitudeSet = false;
    }

    document.getElementById("createGeofenceButton").disabled = !settingsAvailable();
}

function onLongitudeChanged() {
    if (textChangedHandlerDouble(false, "Longitude", longitude)) {
        longitudeSet = true;
    } else {
        longitudeSet = false;
    }

    document.getElementById("createGeofenceButton").disabled = !settingsAvailable();
}

function onRadiusChanged() {
    if (textChangedHandlerDouble(false, "Radius", radius)) {
        radiusSet = true;
    } else {
        radiusSet = false;
    }

    document.getElementById("createGeofenceButton").disabled = !settingsAvailable();
}

function textChangedHandlerDouble(nullAllowed, elementName, e) {
    var valueSet = false;

    var stringValue = new String(e.value);

    var value = decimalFormatter.parseDouble(stringValue);

    if (!value) {
        var msg;
        // value is either empty or alphabetic
        if (0 === stringValue.length) {
            if (!nullAllowed) {
                if (elementName) {
                    msg = elementName + " needs a value";
                    WinJS.log && WinJS.log(msg, "sample", "status");
                }
            }
        } else {
            if (elementName) {
                msg = elementName + " must be a number";
                WinJS.log && WinJS.log(msg, "sample", "status");
            }
        }
    } else {
        valueSet = true;
    }

    if (valueSet) {
        // clear out status message
        WinJS.log && WinJS.log("", "sample", "status");
    }

    return valueSet;
}

function getTimeComponentAsString(str, timeComponent, appendDelimiter) {
    if (timeComponent) {
        if (timeComponent < 10 && str.length) {
            str += "0";
        }

        str += timeComponent.toString();
    } else if (str.length) {
        str += "00";
    }

    if (appendDelimiter && str.length) {
        str += ":";
    }

    return str;
}

function getDurationString(duration) {
    // note that double negation turns a float to an int
    var totalSeconds = ~~(duration / millisecondsPerSecond);
    var days = ~~(totalSeconds / secondsPerDay);
    totalSeconds -= days * secondsPerDay;
    var hours = ~~(totalSeconds / secondsPerHour);
    totalSeconds -= hours * secondsPerHour;
    var minutes = ~~(totalSeconds / secondsPerMinute);
    totalSeconds -= minutes * secondsPerMinute;
    var seconds = totalSeconds;
    var str = new String("");

    str = getTimeComponentAsString(str, days, true);
    str = getTimeComponentAsString(str, hours, true);
    str = getTimeComponentAsString(str, minutes, true);
    str = getTimeComponentAsString(str, seconds, false);

    return str;
}

function refreshControlsFromGeofenceItem(item) {
    if (item) {
        nameElement.value = item.id;
        latitude.value = item.latitude;
        longitude.value = item.longitude;
        radius.value = item.radius;
        geofenceSingleUse.checked = item.singleUse;

        if (0 !== item.dwellTime) {
            dwellTimeField.value = getDurationString(item.dwellTime);
        } else {
            dwellTimeField.value = "";
        }
        if (0 !== item.duration) {
            durationField.value = getDurationString(item.duration);
        } else {
            durationField.value = "";
        }
        if (0 !== item.startTime) {
            startTimeField.value = formatterShortDateLongTime.format(item.startTime);
        } else {
            startTimeField.value = "";
        }

        // Update flags used to enable Create Geofence button
        onNameChanged();
        onLongitudeChanged();
        onLatitudeChanged();
        onRadiusChanged();
    }
}

// are settings available so a geofence can be created?
function settingsAvailable() {
    var fSettingsAvailable = false;

    if (nameSet && latitudeSet && longitudeSet && radiusSet) {
        // also need to test if data is good
        fSettingsAvailable = true;
    }

    return fSettingsAvailable;
}

var TimeSpanFields = Object.freeze({
    day: 0,
    hour: 1,
    minute: 2,
    second: 3
});

function parseTimeSpan(field, defaultValue) {
    var timeSpanValue = 0;
    var vars = field.value.split(':');
    var start = 4 - vars.length;

    if (start >= 0) {
        var loop = 0;
        var varsIndex = start;
        for (; loop < vars.length; loop++, varsIndex++) {
            switch (varsIndex) {
                case TimeSpanFields.day:
                    timeSpanValue += decimalFormatter.parseInt(vars[loop]) * secondsPerDay;
                    break;
                case TimeSpanFields.hour:
                    timeSpanValue += decimalFormatter.parseInt(vars[loop]) * secondsPerHour;
                    break;
                case TimeSpanFields.minute:
                    timeSpanValue += decimalFormatter.parseInt(vars[loop]) * secondsPerMinute;
                    break;
                case TimeSpanFields.second:
                    timeSpanValue += decimalFormatter.parseInt(vars[loop]);
                    break;
            }
        }
    }

    if (0 === timeSpanValue) {
        timeSpanValue = defaultValue;
    }

    timeSpanValue *= millisecondsPerSecond;

    return timeSpanValue;
}

function generateGeofence() {
    var geofence = null;
    try {
        var fenceKey = nameElement.value;

        var position = {
            latitude: decimalFormatter.parseDouble(latitude.value),
            longitude: decimalFormatter.parseDouble(longitude.value),
            altitude: 0
        };
        var radiusValue = decimalFormatter.parseDouble(radius.value);

        // the geofence is a circular region
        var geocircle = new Windows.Devices.Geolocation.Geocircle(position, radiusValue);

        var singleUse = false;

        if (geofenceSingleUse.checked) {
            singleUse = true;
        }

        // want to listen for enter geofence, exit geofence and remove geofence events
        var mask = 0;

        mask = mask | Windows.Devices.Geolocation.Geofencing.MonitoredGeofenceStates.entered;
        mask = mask | Windows.Devices.Geolocation.Geofencing.MonitoredGeofenceStates.exited;
        mask = mask | Windows.Devices.Geolocation.Geofencing.MonitoredGeofenceStates.removed;

        var dwellTimeSpan = new Number(parseTimeSpan(dwellTimeField, defaultDwellTimeSeconds));
        var durationTimeSpan = null;
        if (durationField.value.length) {
            durationTimeSpan = new Number(parseTimeSpan(durationField, 0));
        } else {
            durationTimeSpan = new Number(0); // duration needs to be set since start time is set below
        }
        var startDateTime = null;
        if (startTimeField.value.length) {
            startDateTime = new Date(startTimeField.value);
        } else {
            startDateTime = new Date(); // if you don't set start time in JavaScript the start time defaults to 1/1/1601
        }

        geofence = new Windows.Devices.Geolocation.Geofencing.Geofence(fenceKey, geocircle, mask, singleUse, dwellTimeSpan, startDateTime, durationTimeSpan);
    } catch (ex) {
        WinJS.log && WinJS.log(ex.toString(), "sample", "error");
    }

    return geofence;
}
